src/coreclr/vm/interoplibinterface_comwrappers.cpp

```diff
@@ -672,1402 +48,682 @@ namespace
-
-            STRESS_LOG2(LF_INTEROP, LL_INFO100, "Created Target for External: 0x%p => 0x%p\n", OBJECTREFToObject(gc.objRef), *trackerTarget);
-            GCPROTECT_END();
-        }
-        END_EXTERNAL_ENTRYPOINT;
-
-        return hr;
-    }
-
-    TryInvokeICustomQueryInterfaceResult TryInvokeICustomQueryInterface(
-        _In_ InteropLib::OBJECTHANDLE handle,
-        _In_ REFGUID iid,
-        _Outptr_result_maybenull_ void** obj) noexcept
-    {
-        CONTRACTL
-        {
-            NOTHROW;
-            MODE_ANY;
-            PRECONDITION(handle != NULL);
-            PRECONDITION(obj != NULL);
-        }
-        CONTRACTL_END;
-
-        *obj = NULL;
-
-        // If this is a GC thread, then someone is trying to query for something
-        // at a time when we can't run managed code.
-        if (IsGCThread())
-            return TryInvokeICustomQueryInterfaceResult::OnGCThread;
-
-        // Ideally the BEGIN_EXTERNAL_ENTRYPOINT/END_EXTERNAL_ENTRYPOINT pairs
-        // would be used here. However, this code path can be entered from within
-        // and from outside the runtime.
-        MAKE_CURRENT_THREAD_AVAILABLE_EX(GetThreadNULLOk());
-        if (CURRENT_THREAD == NULL)
-        {
-            CURRENT_THREAD = SetupThreadNoThrow();
-
-            // If we failed to set up a new thread, we are going to indicate
-            // there was a general failure to invoke instead of failing fast.
-            if (CURRENT_THREAD == NULL)
-                return TryInvokeICustomQueryInterfaceResult::FailedToInvoke;
-        }
-
-        HRESULT hr;
-        auto result = TryInvokeICustomQueryInterfaceResult::FailedToInvoke;
-        EX_TRY_THREAD(CURRENT_THREAD)
-        {
-            // Switch to Cooperative mode since object references
-            // are being manipulated.
-            GCX_COOP();
+    QCALL_CONTRACT;
 
-            struct
-            {
-                OBJECTREF objRef;
-            } gc;
-            gc.objRef = NULL;
-            GCPROTECT_BEGIN(gc);
+    _ASSERTE(fpQueryInterface != NULL);
+    _ASSERTE(fpAddRef != NULL);
+    _ASSERTE(fpRelease != NULL);
 
-            // Get the target of the external object's reference.
-            ::OBJECTHANDLE objectHandle = static_cast<::OBJECTHANDLE>(handle);
-            gc.objRef = ObjectFromHandle(objectHandle);
+    BEGIN_QCALL;
 
-            result = (TryInvokeICustomQueryInterfaceResult)CallICustomQueryInterface(&gc.objRef, iid, obj);
+    InteropLib::Com::GetIUnknownImpl(fpQueryInterface, fpAddRef, fpRelease);
 
-            GCPROTECT_END();
-        }
-        EX_CATCH_HRESULT(hr);
+    END_QCALL;
+}
 
-        // Assert valid value.
-        _ASSERTE(TryInvokeICustomQueryInterfaceResult::Min <= result
-            && result <= TryInvokeICustomQueryInterfaceResult::Max);
+void ComWrappersNative::DestroyExternalComObjectContext(_In_ void* contextRaw)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_ANY;
```

Comment:
Is this the correct mode? When is this API called?

Reply:
Yes, the caller and its caller are both `MODE_ANY`.

We can likely change it, but that would mean we'd need to trace back all syncblock cleanup and update all of the functions in the call tree.